### PR TITLE
(PUP-9258) Refactor loading of PAL

### PR DESCRIPTION
**Pull request by hlindberg**
This refactors PAL (without any need for users to change anything):

* the implementation is now in multiple files under `puppet/pal`
* the `puppet_pal` file now just requires puppet and the pal api
* puppet now also requires the PAL api

While it would be possible to drop the loading of `puppet/pal/pal_api`
because `puppet` requires it the longer term intention is that `pal`
will not require everything that puppet requires.

### CODE REVIEW

File: lib/puppet/pal/catalog_compiler.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    # Can be called multiple times.
+    #
+    def evaluate_additions
+      internal_compiler.evaluate_additions
+    end
+
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/catalog_compiler.rb]: missing newline

File: lib/puppet/pal/compiler.rb [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+    private
+
+    def topscope
+      internal_compiler.topscope
+    end
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/compiler.rb]: missing newline

File: lib/puppet/pal/function_signature.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    # @api public
+    #
+    def callables
+      signatures = @func.dispatcher.to_type
+      signatures.is_a?(Puppet::Pops::Types::PVariantType) ? signatures.types : [signatures]
+    end
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/function_signature.rb]: missing newline

File: lib/puppet/pal/plan_signature.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    # @api public
+    #
+    def params_type
+      dispatcher = @plan_func.class.dispatcher.dispatchers[0]
+      dispatcher.params_struct
+    end
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/plan_signature.rb]: missing newline

File: lib/puppet/pal/script_compiler.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    # @param error_collector [Array<Puppet::DataTypes::Error>] an optional array that will receive errors during load
+    # @return [Array<Puppet::Pops::Loader::TypedName>] an array of typed names
+    #
+    def list_tasks(filter_regex = nil, error_collector = nil)
+      list_loadable_kind(:task, filter_regex, error_collector)
+    end
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/script_compiler.rb]: missing newline

File: lib/puppet/pal/task_signature.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    # the task such as the description, parameters, output, etc.
+    #
+    # @return [Puppet::Pops::Types::PuppetObject] An instance of a dynamically created Task class
+    def task
+      @task
+    end
+  end
+
+end
+end
```
Comment (thallgren) [inline lib/puppet/pal/task_signature.rb]: missing newline

**Comment by hlindberg** [comment]
Meh - EOF's fixed. Waiting for github to get their act together - still no hooks for tests etc.

**Pull request merged**